logging: Ensure -debug=0/none behaves consistently with -nodebug

danielabrozzoni opened the pull request:
Previously, -nodebug cleared all prior -debug configurations in the command line while allowing subsequent debug options to be applied.
However, -debug=0 and -debug=none completely disabled debugging, even for categories specified afterward.
    
This commit ensures consistency by making -debug=0 and -debug=none behave like -nodebug: they now clear previously set debug configurations but do not disable debugging for categories specified later.

See https://github.com/bitcoin/bitcoin/pull/30529#discussion_r1930956563

danielabrozzoni commented on doc/release-notes-31767.md:
```diff
@@ -0,0 +1,3 @@
+RPC
```
You're right, thanks! Updated with "Logging" as it feels more appropriate to me too

luke-jr replied:
PR title and commit message too, please

hodlinator commented on src/init/common.cpp:
```diff
@@ -80,15 +80,21 @@ util::Result<void> SetLoggingLevel(const ArgsManager& args)
-                    return util::Error{strprintf(_("Unsupported logging category %s=%s."), "-debug", cat)};
-                }
+        // Special-case: Disregard any debugging categories appearing before -debug=0/none
+        auto last_negated = categories.rbegin();
+        for (; last_negated != categories.rend(); ++last_negated) {
+            if (*last_negated == "0" || *last_negated == "none") {
+                break;
+            }
+        }
```
nit: I tried to replicate what [ChatGPT suggested to ryanofsky](https://github.com/bitcoin/bitcoin/pull/30529#discussion_r1930956563), but `std::views` wouldn't compile (didn't look into modifying my setup). Was able to combine it with your version though into:
```suggestion
        auto last_negated = std::find_if(categories.rbegin(), categories.rend(),
            [](const std::string& cat) { return cat == "0" || cat == "none"; });
```

hodlinator commented on src/init/common.cpp:
```diff
@@ -80,15 +80,21 @@ util::Result<void> SetLoggingLevel(const ArgsManager& args)
+            }
+        }
+
+        auto range_to_process = (last_negated == categories.rend()) ? categories : std::ranges::subrange(last_negated.base(), categories.end());
+
+        for (const auto& cat : range_to_process) {
+            if (!LogInstance().EnableCategory(cat)) {
+                return util::Error{strprintf(_("Unsupported logging category %s=%s."), "-debug", cat)};
```
mega-nit: I find this marginally nicer, but still not great:
```suggestion
        for (auto it = last_negated == categories.rend() ? categories.begin() : last_negated.base();
            it != categories.end(); ++it) {
            if (!LogInstance().EnableCategory(*it)) {
                return util::Error{strprintf(_("Unsupported logging category %s=%s."), "-debug", *it)};
```

hodlinator commented on test/functional/feature_logging.py:
```diff
@@ -99,6 +99,18 @@ def run_test(self):
+        disable_debug_options = [
+            '-debug=0',
+            '-debug=none',
+            '-nodebug'
```
nit: I tested breaking out an initial commit which only adds the test for `-nodebug` to prove existing behavior, before the commit performing the C++ change and adding `0` + `none` tests (it passed). You might want to rewrite the commit history to do that as well, but I'm okay with the current approach.

hodlinator commented on test/functional/feature_logging.py:
```diff
@@ -99,6 +99,18 @@ def run_test(self):
+        ]
+
+        for debug_opt in disable_debug_options:
+            self.restart_node(0, ['-debug=http', debug_opt, '-debug=rpc'])
+            logging = self.nodes[0].logging()
+            assert logging['rpc']
```
nit: Could add invalid `abc` category prior to disabling, and verify that *multiple* categories after the disabling work.
```suggestion
            self.restart_node(0, ['-debug=http', '-debug=abc', debug_opt, '-debug=rpc', '-debug=net'])
            logging = self.nodes[0].logging()
            assert logging['rpc']
            assert logging['net']
```

hodlinator replied:
Nice touch on adding a check for non-existence of abc!
https://github.com/bitcoin/bitcoin/blob/bb8309723cdd3ce0378fbe9708bd600dfaa157aa/test/functional/feature_logging.py#L113

hodlinator reviewed (approved):
ACK 1f4aff75c308226d58a45460982f75287e15cb81

### Concept

Having `-debug=0` and `-debug=none` perform the same effect as `-nodebug` in discarding all preceding categories makes more sense, instead of disabling all categories including the ones following them.

### Testing

As well as running the new test as-is, I also verified stated behavior of `-nodebug` prior to the C++ change by modifying included test.

### Suggestions

Only found nits, the `last_negated` calculation in particular feels unnecessarily verbose & mutation heavy.

danielabrozzoni committed:
test: `-nodebug` clears previously set debug options

danielabrozzoni committed:
logging: Ensure -debug=0/none behaves consistently with -nodebug

Previously, -nodebug cleared all prior -debug configurations in the
command line while allowing subsequent debug options to be applied.
However, -debug=0 and -debug=none completely disabled debugging,
even for categories specified afterward.

This commit ensures consistency by making -debug=0 and -debug=none
behave like -nodebug: they now clear previously set debug configurations
but do not disable debugging for categories specified later.

Co-Authored-By: Ryan Ofsky <ryan@ofsky.org>

danielabrozzoni committed:
test: `-debug=0` and `-debug=none` behave similarly to `-nodebug`

danielabrozzoni reviewed (commented):
Thank you for the reviews! In my last push:
- I removed every occurences of "rpc" (sorry I didn't catch them the first time!)
- I've renamed a couple of variables, added a comment, and addressed @hodlinator's nits, to improve clarity

hodlinator commented on src/init/common.cpp:
```diff
@@ -80,15 +80,23 @@ util::Result<void> SetLoggingLevel(const ArgsManager& args)
+        auto last_negated = std::find_if(categories.rbegin(), categories.rend(),
+            [](const std::string& cat) { return cat == "0" || cat == "none"; });
+
+        for (; last_negated != categories.rend(); ++last_negated) {
+            if (*last_negated == "0" || *last_negated == "none") {
+                break;
+            }
+        }
+
```
Seems like you forgot to remove this? (`last_negated` should theoretically be able to be marked `const`, but makes line go over 80 chars).
```suggestion
```

danielabrozzoni replied:
Of course! Thank you, updated, and made last_negated and categories_to_process const

hodlinator reviewed (approved):
re-ACK 7afeaa24693039730c9ae00c325c2b9d0e2deda1

Changes since [previous ACK](https://github.com/bitcoin/bitcoin/pull/31767#pullrequestreview-2599809742):
- Broke out initial commit which tests prior behavior (with minor improvements).
- Terser test for last occurrence of negated arg suggested by self.

ryanofsky reviewed (approved):
Code review ACK 7afeaa24693039730c9ae00c325c2b9d0e2deda1. Nicely implemented change with test and release notes, and I like how the test is implemented as the first commit.

maflcko reviewed (approved):
left two nits

review ACK 7afeaa24693039730c9ae00c325c2b9d0e2deda1 👡

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 7afeaa24693039730c9ae00c325c2b9d0e2deda1 👡
ArIz38cfPYwJHMq9d1QKT9t9zk+QWETW8C/1e1B83bsin8cm2QVQRHActkHAgZSX49gcc2TtDyMFdPVuDcWZDw==
```

</details>

ryanofsky commented on src/init/common.cpp:
```diff
@@ -80,15 +80,17 @@ util::Result<void> SetLoggingLevel(const ArgsManager& args)
 util::Result<void> SetLoggingCategories(const ArgsManager& args)
 {
     if (args.IsArgSet("-debug")) {
```
re: https://github.com/bitcoin/bitcoin/pull/31767#discussion_r1954308137

> Nit in [a8fedb3](https://github.com/bitcoin/bitcoin/commit/a8fedb36a71ac193fc158284b14d4eb474e5ab62): Why is this needed? It seems like a no-op:

Right it is is a no-op, see #30529 which makes the change you suggested and removes many other misuses of IsArgSet

ryanofsky replied:
> Thx for confirming. It would be nice to remove it here, so that it doesn't have to be touched (and reviewed) again in another pull request. 

Yep, #30529 does have 2 acks though so getting close
